File: quantum-safe/hybrid_kem.go
package quantumsafe

import (
	"crypto/sha256"

	"github.com/cloudflare/circl/dh/x25519"
	"github.com/cloudflare/circl/kem/mlkem/mlkem768"
)

// HybridKEM combines X25519 ECDH with ML-KEM-768 encapsulation (X25519MLKEM768).
func HybridKEM() (bool, error) {
	// classical half: X25519 ECDH between two parties
	var aPriv, aPub, bPriv, bPub x25519.Key
	x25519.KeyGen(&aPub, &aPriv)
	x25519.KeyGen(&bPub, &bPriv)
	var ssClassicalA, ssClassicalB x25519.Key
	x25519.Shared(&ssClassicalA, &aPriv, &bPub)
	x25519.Shared(&ssClassicalB, &bPriv, &aPub)

	// post-quantum half: ML-KEM-768
	scheme := mlkem768.Scheme()
	pk, sk, err := scheme.GenerateKeyPair()
	if err != nil {
		return false, err
	}
	ct, ssPqA, err := scheme.Encapsulate(pk)
	if err != nil {
		return false, err
	}
	ssPqB, err := scheme.Decapsulate(sk, ct)
	if err != nil {
		return false, err
	}

	// combine each side: SHA-256(ssPq || ssClassical)
	sender := sha256.Sum256(append(ssPqA, ssClassicalA[:]...))
	recipient := sha256.Sum256(append(ssPqB, ssClassicalB[:]...))
	return sender == recipient, nil
}

File: quantum-safe/hybridKem.js
// Hybrid PQC KEM: X25519 + ML-KEM-768 (X25519MLKEM768).
// npm i @noble/post-quantum @noble/curves @noble/hashes
'use strict';
const { ml_kem768 } = require('@noble/post-quantum/ml-kem');
const { x25519 } = require('@noble/curves/ed25519');
const { sha256 } = require('@noble/hashes/sha256');

function hybridKem() {
  // classical half: X25519 ECDH
  const aPriv = x25519.utils.randomPrivateKey();
  const bPriv = x25519.utils.randomPrivateKey();
  const aPub = x25519.getPublicKey(aPriv);
  const bPub = x25519.getPublicKey(bPriv);
  const ssClassicalA = x25519.getSharedSecret(aPriv, bPub);
  const ssClassicalB = x25519.getSharedSecret(bPriv, aPub);

  // post-quantum half: ML-KEM-768
  const { publicKey, secretKey } = ml_kem768.keygen();
  const { cipherText, sharedSecret } = ml_kem768.encapsulate(publicKey);
  const ssPqB = ml_kem768.decapsulate(cipherText, secretKey);

  // combine: SHA-256(ss_pq || ss_classical)
  const sender = sha256(Buffer.concat([Buffer.from(sharedSecret), Buffer.from(ssClassicalA)]));
  const recipient = sha256(Buffer.concat([Buffer.from(ssPqB), Buffer.from(ssClassicalB)]));
  return Buffer.from(sender).equals(Buffer.from(recipient));
}

console.log('X25519 + ML-KEM-768 hybrid secret match:', hybridKem());
module.exports = { hybridKem };
